internal/downlink/data/data.go

```diff
@@ -387,14 +391,177 @@ func setTXParameters(ctx *dataContext) error {
 	return nil
 }
 
-func setDataTXInfo(ctx *dataContext) error {
+func sfToSNR(sf int) float64 {
+	// Minimal SNR for SFs:
+	// SF 7:   -7.5
+	// SF 8:   -10
+	// SF 9:   -12.5
+	// SF 10:  -15
+	// SF 11:  -17.5
+	// SF 12:  -20
+	switch sf {
+	case 12:
+		return -20
+	case 11:
+		return -17.5
+	case 10:
+		return -15
+	case 9:
+		return -12.5
+	case 8:
+		return -10
+	case 7:
+		return -7.5
+	default:
+		return 0
+	}
+}
+
+func isRX2Only(ctx *dataContext) (b bool, err error) {
+	// do not block RX1 until RX Settings of node are set correctly
+	if ctx.DeviceSession.RX2Frequency != rx2Frequency || ctx.DeviceSession.RX2DR != uint8(rx2DR) ||
+		ctx.DeviceSession.RX1DROffset != uint8(rx1DROffset) || ctx.DeviceSession.RXDelay != uint8(rx1Delay) {
+		return false, nil
+	}
+
+	// get uplink DR
+	uplinkDR, err := helpers.GetDataRateIndex(true, ctx.RXPacket.TXInfo, band.Band())
+	if err != nil {
+		return true, errors.Wrap(err, "get data-rate index error")
+	}
+
+	// get rx1 data-rate
+	drRX1Index, err := band.Band().GetRX1DataRateIndex(uplinkDR, int(ctx.DeviceSession.RX1DROffset))
+	if err != nil {
+		return true, errors.Wrap(err, "get rx1 data-rate index error")
+	}
+
+	if rx2OnlyDR >= 0 { // if the limit is set via config file, apply the settings accordingly
+		if rx2OnlyDR >= drRX1Index {
+			return true, nil
+		} else {
+			return false, nil
+		}
+	} else if rx2OnlyDR == -1 { // try to decide based on simplified link budget
+		var (
+			t			  int
+			txPowerRX1    int32
+			txPowerRX2    int32
+			freqRX1       uint32
+			freqRX2       uint32
+			linkBudgetRX1 float64
+			linkBudgetRX2 float64
+			drRX1		  loraband.DataRate
+			drRX2		  loraband.DataRate
+			logBWRX1 	  float64
+			logBWRX2	  float64
+		)
+
+		// get rx1 data-rate
+		drRX1, err = band.Band().GetDataRate(drRX1Index)
+		if err != nil {
+			return true, err // on error, omit RX1
+		}
+
+		if drRX1.Modulation == loraband.FSKModulation { //never block RX1 for FSK
+			return false, nil
+		}
+
+		// get rx2 data-rate
+		drRX2, err = band.Band().GetDataRate(int(ctx.DeviceSession.RX2DR))
+		if err != nil {
+			return false, err // on error, don't block RX1
+		}
+
+		// get RX1 and RX2 freq
+		t, err = band.Band().GetRX1FrequencyForUplinkFrequency(int(ctx.RXPacket.TXInfo.Frequency))
+		if err != nil {
+			return true, err //could not get RX1 Frequency for any reason, so just send RX2 then
+		} else {
+			freqRX1 = uint32(t)
+		}
+
+		freqRX2 = uint32(ctx.DeviceSession.RX2Frequency)
+
+		// get RX1 and RX2 TX Power
+		if downlinkTXPower != -1 {
+			txPowerRX1 = int32(downlinkTXPower)
+			txPowerRX2 = int32(downlinkTXPower)
+		} else {
+			txPowerRX1 = int32(band.Band().GetDownlinkTXPower(int(freqRX1)))
+			txPowerRX2 = int32(band.Band().GetDownlinkTXPower(int(freqRX2)))
+		}
+
+		if drRX1.Bandwidth == 125 {
+			logBWRX1 = 21 // 10*log10(125)
+		} else {
+			logBWRX1 = 24 // 10*log10(250)
```

Comment:
I think it is better to calculate the actual log, as this might as well be 500 kHz. Even when all logs are pre-calculated, it could cause issues in the future as new bandwidth values might be added in the future, causing the code to make assumptions that are no longer true. See:

https://golang.org/pkg/math/#Log10

Reply:
Allright, I come from the embedded side but I guess log is quite cheap on modern systems even for scalable systems/huge amount of traffic.